### PR TITLE
Rotary-like control for angles

### LINKED ISSUE #919: Need a rotary control for angles
We need a simple, small-ish rotary control with which the user can more intuitively enter an angle. The use case initially is the rotation parameter of the component transformation.

### PR DESCRIPTION
**Pull request by fatih-erikli**
Fixes #919.

### CODE REVIEW

File: src/fontra/client/web-components/rotary-control.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    );
+    this.overlay = overlay;
+    this.shadowRoot.appendChild(overlay);
+  }
+
+  render() {
+    return html.div({ class: "rotary-control" }, [
+      (this.knob = html.div(
+        {
+          onwheel: (event) => {
```
Comment (justvanrossum) [inline src/fontra/client/web-components/rotary-control.js:109]: Unless we can support dragBegin/dragEnd for wheel events we should drop onwheel.

File: src/fontra/client/web-components/ui-form.js [hunk 2]
```diff
@@ -171,6 +172,37 @@ export class Form extends SimpleElement {
     valueElement.appendChild(inputElement);
   }
 
+  _addEditAngle(valueElement, fieldItem) {
+    const inputElement = document.createElement("input");
```
Comment (justvanrossum) [inline src/fontra/client/web-components/ui-form.js:176]: We have an "input" shortcut in html-utils.js. I'm pretty sure you can write this instead:
```js
    const inputElement = html.input({type: "number", "value": ...etc...});
```

File: src/fontra/client/web-components/ui-form.js [hunk 3]
```diff
@@ -171,6 +172,37 @@ export class Form extends SimpleElement {
...
+    inputElement.step = "any";
+
+    inputElement.disabled = fieldItem.disabled;
+    inputElement.onchange = (event) => {
+      let value = parseFloat(inputElement.value);
+      this._fieldChanging(fieldItem.key, value);
+      rotaryControl.value = value;
+    };
+
+    const rotaryControl = html.createDomElement("rotary-control", {
```
[the remark this answers is not in the document]
Comment (justvanrossum) [reply src/fontra/client/web-components/ui-form.js:189]: Ok, fair enough!

**Comment by justvanrossum** [comment]
Please rebase this.

**Review by justvanrossum — COMMENTED** [review]
- When setting the angle numerically to (say) -50, the thumb get shown rotated counter-clockwise, but when dragging the thumb, the value changes in the opposite direction
- Dragging the thumb is often jumpy
- Dragging to the right rotates clockwise, and draging up rotates counter-clockwise. I think up should also rotate clockwise.

**Comment by fatih-erikli** [comment]
> When setting the angle numerically to (say) -50, the thumb get shown rotated counter-clockwise, but when dragging the thumb, the value changes in the opposite direction

This is fixed. The rotation of glyph on the canvas is applied counterclockwise. There is conversion to the opposite side happening when dragging the thumb.

**Comment by fatih-erikli** [comment]
Maybe I delete the conversion and keep it as it is? The rotation will be applied to the opposite side when dragging the thumb, is that ok? @justvanrossum.

**Comment by justvanrossum** [comment]
> Maybe I delete the conversion and keep it as it is? The rotation will be applied to the opposite side when dragging the thumb, is that ok?

You mean rotate the rotary control clockwise while the component rotates counter-clockwise? No, they should match, that's the whole point.

Btw. please remove that demo panel, and just test with it with a selected component.

The movement is still very jumpy:

https://github.com/googlefonts/fontra/assets/4246121/2560e266-5b82-4ad2-bcf3-68df7f12358f

**Comment by fatih-erikli** [comment]
Can you test the jumpy behavior with the latest commit?

**Comment by fatih-erikli** [comment]
Hmm no this is a different kind of jumpy behavior. Video has just been loaded.

**Comment by fatih-erikli** [comment]
Hmm, it looks like it's fixed with the latest commit.

**Comment by justvanrossum** [comment]
It still jumps if you move across the NW - SE axis:

image

**Comment by fatih-erikli** [comment]
This is tricky. I tried to fix that by caching the rotation axis, but it expects the drag to be same with when the movement started until the mouse up.

File: src/fontra/views/editor/editor.js [hunk 4]
```diff
@@ -62,6 +62,7 @@ import { themeController } from "/core/theme-settings.js";
 import { MenuItemDivider, showMenu } from "/web-components/menu-panel.js";
 import { dialog, dialogSetup } from "/web-components/modal-dialog.js";
 import { parsePluginBasePath } from "/web-components/plugin-manager.js";
+import { RotaryControl } from "/web-components/rotary-control.js";
```
Comment (justvanrossum) [inline src/fontra/views/editor/editor.js:65]: I think this import belongs in ui-form, where it's used via createDomElement.

**Comment by justvanrossum** [comment]
Can you make the rotary control as big as possible without affecting the item spacing? It is currently smaller than the numeric input. Can it be the same size?

**Comment by fatih-erikli** [comment]
Maximum height without affecting the item spacing is the current one. There was a margin, I deleted it, they should seem in same height now.

**Pull request merged**